File: noise.lua
--------------------------------------------------------------------------------
-- NOISE GENERATOR
--------------------------------------------------------------------------------

-- knobs
noise_white = true

-- internal
-- Trammel Stochastic Algorithm
-- https://web.archive.org/web/20160705171002/http://home.earthlink.net/~ltrammell/tech/pinkalg.htm
local function noise_pink_band(amplitude, probability)
    return coroutine.wrap(
        function()
            while true do
                local sample = amplitude * math.random(-1, 1)
                repeat
                    local random = coroutine.yield(sample)
                until random > probability
            end
        end
    )
end
noise_pink_amp = 4.6306e-3 + 5.9961e-3 + 8.3586e-3
noise_pink_band_1 = noise_pink_band(4.6306e-3 / noise_pink_amp, 3.1878e-1)
noise_pink_band_2 = noise_pink_band(5.9961e-3 / noise_pink_amp, 7.7686e-1)
noise_pink_band_3 = noise_pink_band(8.3586e-3 / noise_pink_amp, 9.7785e-1)

-- out
noise_out = 0

function update_noise()
    if noise_white then
        noise_out = math.random(-1, 1)
    else
        local random = math.random()
        noise_out = noise_pink_band_1(random) + noise_pink_band_2(random) + noise_pink_band_3(random)
    end
end

File: noise.cpp
////////////////////////////////////////////////////////////////////////////////
// NOISE GENERATOR
////////////////////////////////////////////////////////////////////////////////

#include <cmath>

#include "synth.hpp"

// Trammel Stochastic Algorithm
// https://web.archive.org/web/20160705171002/http://home.earthlink.net/~ltrammell/tech/pinkalg.htm
static float amplitudes[] = {
    0.24390449452997848,
    0.3158285620980443,
    0.4402669433719773
};
static float probabilities[] = {
    3.1878e-1,
    7.7686e-1,
    9.7785e-1
};
static float pink_band(int* state, float amplitude, float probability, float sample, float random) {
    crBegin(*state);
    while (true) {
        sample = amplitude * (random * 2 -1);
        do {
            crReturn(*state, sample);
        } while (random < probability);
    }
    crFinish();

    abort();
}

void Synth::update_noise() {
    auto random = rand() / float(RAND_MAX);
    if (noise.white) {
        noise.out = random * 2 - 1;
    } else {
        noise.out = 0;
        for (int i = 0; i < 3; ++i) {
            noise.out += noise.samples[i] = pink_band(
                &noise.state[i],
                amplitudes[i],
                probabilities[i],
                noise.samples[i],
                random
            );
        }
    }
}
